File: services/saas-api/src/main/java/com/yunyan/saasapi/domain/permission/PermissionCodes.java
package com.yunyan.saasapi.domain.permission;

/** Stable permission codes seeded in V6__permissions.sql; used by admin API and @PreAuthorize (D-02+). */
public final class PermissionCodes {

  public static final String ADMIN_TENANTS_READ = "admin:tenants:read";
  public static final String ADMIN_TENANTS_WRITE = "admin:tenants:write";
  public static final String ADMIN_USERS_READ = "admin:users:read";
  public static final String ADMIN_USERS_WRITE = "admin:users:write";
  public static final String ADMIN_ROLES_READ = "admin:roles:read";
  public static final String ADMIN_ROLES_WRITE = "admin:roles:write";
  public static final String ADMIN_MEMBERS_READ = "admin:members:read";
  public static final String ADMIN_MEMBERS_WRITE = "admin:members:write";
  public static final String WORKSPACE_USE = "workspace:use";
  public static final String WORKSPACE_MAP_READ = "workspace:map:read";
  public static final String WORKSPACE_MAP_WRITE = "workspace:map:write";

  public static final String BILLING_WALLET_READ = "billing:wallet:read";
  public static final String BILLING_LEDGER_READ = "billing:ledger:read";
  public static final String BILLING_RECHARGE_CREATE = "billing:recharge:create";
  public static final String BILLING_USAGE_READ = "billing:usage:read";
  public static final String BILLING_TRANSFER_CREATE = "billing:transfer:create";
  public static final String ADMIN_BILLING_READ = "admin:billing:read";
  public static final String ADMIN_BILLING_ADJUST = "admin:billing:adjust";
  public static final String ADMIN_BILLING_PACKAGES_WRITE = "admin:billing:packages:write";
  public static final String ADMIN_BILLING_REFUND = "admin:billing:refund";
  public static final String ADMIN_AUDIT_READ = "admin:audit:read";
  public static final String ADMIN_AUDIT_EXPORT = "admin:audit:export";
  public static final String ADMIN_MENUS_READ = "admin:menus:read";
  public static final String ADMIN_MENUS_WRITE = "admin:menus:write";
  public static final String ADMIN_IMPERSONATE = "admin:impersonate";

  /** Requires explicit audit permission (no tenants:read fallback). */
  public static final String ADMIN_AUDIT_READ_AUTHORITIES =
      "hasAuthority('" + ADMIN_AUDIT_READ + "')";

  public static final String ADMIN_AUDIT_EXPORT_AUTHORITIES =
      "hasAuthority('" + ADMIN_AUDIT_EXPORT + "')";

  /** Platform-scoped permissions; any one grants access to `/v1/admin/**` (D-02). */
  public static final String[] PLATFORM_ADMIN_AUTHORITIES = {
    ADMIN_TENANTS_READ,
    ADMIN_TENANTS_WRITE,
    ADMIN_USERS_READ,
    ADMIN_USERS_WRITE,
    ADMIN_ROLES_READ,
    ADMIN_ROLES_WRITE,
    ADMIN_AUDIT_READ,
    ADMIN_AUDIT_EXPORT,
    ADMIN_MENUS_READ,
    ADMIN_MENUS_WRITE,
    ADMIN_IMPERSONATE,
    ADMIN_BILLING_READ,
    ADMIN_BILLING_ADJUST,
    ADMIN_BILLING_PACKAGES_WRITE,
    ADMIN_BILLING_REFUND
  };

  /** Tenant-scoped permissions for `/v1/admin/**` member routes (D-06). */
  public static final String[] TENANT_MEMBER_ADMIN_AUTHORITIES = {
    ADMIN_MEMBERS_READ,
    ADMIN_MEMBERS_WRITE
  };

  /** Any authority that may enter `/v1/admin/**` (except permitAll ping). */
  public static final String[] ADMIN_GATE_AUTHORITIES = {
    ADMIN_TENANTS_READ,
    ADMIN_TENANTS_WRITE,
    ADMIN_USERS_READ,
    ADMIN_USERS_WRITE,
    ADMIN_ROLES_READ,
    ADMIN_ROLES_WRITE,
    ADMIN_AUDIT_READ,
    ADMIN_AUDIT_EXPORT,
    ADMIN_MENUS_READ,
    ADMIN_MENUS_WRITE,
    ADMIN_IMPERSONATE,
    ADMIN_MEMBERS_READ,
    ADMIN_MEMBERS_WRITE
  };

  private PermissionCodes() {}
}

File: packages/auth/src/permission-codes.ts
import { SaaSRole } from './types'

/** 与 saas-api V6__permissions.sql / PermissionCodes.java 对齐 */
export const PermissionCodes = {
  ADMIN_TENANTS_READ: 'admin:tenants:read',
  ADMIN_TENANTS_WRITE: 'admin:tenants:write',
  ADMIN_USERS_READ: 'admin:users:read',
  ADMIN_USERS_WRITE: 'admin:users:write',
  ADMIN_ROLES_READ: 'admin:roles:read',
  ADMIN_ROLES_WRITE: 'admin:roles:write',
  ADMIN_MEMBERS_READ: 'admin:members:read',
  ADMIN_MEMBERS_WRITE: 'admin:members:write',
  WORKSPACE_USE: 'workspace:use',
  WORKSPACE_MAP_READ: 'workspace:map:read',
  WORKSPACE_MAP_WRITE: 'workspace:map:write',
  BILLING_WALLET_READ: 'billing:wallet:read',
  BILLING_LEDGER_READ: 'billing:ledger:read',
  BILLING_RECHARGE_CREATE: 'billing:recharge:create',
  BILLING_USAGE_READ: 'billing:usage:read',
  ADMIN_BILLING_READ: 'admin:billing:read',
  ADMIN_BILLING_ADJUST: 'admin:billing:adjust',
  ADMIN_BILLING_PACKAGES_WRITE: 'admin:billing:packages:write',
} as const

export type PermissionCode = (typeof PermissionCodes)[keyof typeof PermissionCodes]

/** 种子角色默认权限（联调 / mock；运行时以 JWT / users/me 为准） */
export const ROLE_DEFAULT_PERMISSIONS = {
  [SaaSRole.PLATFORM_ADMIN]: [
    PermissionCodes.ADMIN_TENANTS_READ,
    PermissionCodes.ADMIN_TENANTS_WRITE,
    PermissionCodes.ADMIN_USERS_READ,
    PermissionCodes.ADMIN_USERS_WRITE,
    PermissionCodes.ADMIN_ROLES_READ,
    PermissionCodes.ADMIN_ROLES_WRITE,
    PermissionCodes.ADMIN_MEMBERS_READ,
    PermissionCodes.ADMIN_MEMBERS_WRITE,
    PermissionCodes.ADMIN_BILLING_READ,
    PermissionCodes.ADMIN_BILLING_ADJUST,
    PermissionCodes.ADMIN_BILLING_PACKAGES_WRITE,
  ],
  [SaaSRole.TENANT_ADMIN]: [
    PermissionCodes.ADMIN_MEMBERS_READ,
    PermissionCodes.ADMIN_MEMBERS_WRITE,
    PermissionCodes.WORKSPACE_USE,
    PermissionCodes.WORKSPACE_MAP_READ,
    PermissionCodes.WORKSPACE_MAP_WRITE,
    PermissionCodes.BILLING_WALLET_READ,
    PermissionCodes.BILLING_LEDGER_READ,
    PermissionCodes.BILLING_RECHARGE_CREATE,
    PermissionCodes.BILLING_USAGE_READ,
  ],
  [SaaSRole.MEMBER]: [
    PermissionCodes.WORKSPACE_USE,
    PermissionCodes.WORKSPACE_MAP_READ,
    PermissionCodes.WORKSPACE_MAP_WRITE,
    PermissionCodes.BILLING_WALLET_READ,
    PermissionCodes.BILLING_LEDGER_READ,
    PermissionCodes.BILLING_RECHARGE_CREATE,
  ],
  [SaaSRole.VIEWER]: [
    PermissionCodes.WORKSPACE_USE,
    PermissionCodes.WORKSPACE_MAP_READ,
    PermissionCodes.BILLING_WALLET_READ,
    PermissionCodes.BILLING_LEDGER_READ,
  ],
} as const satisfies Record<SaaSRole, readonly PermissionCode[]>
